### PR TITLE
ConDec-502: Extract decision knowledge elements from feature branch changed files

### PR DESCRIPTION
**Pull request by lw2011**
Extracts rationale from (added, modified, deleted) java code files of a feature branch

### CODE REVIEW

File: src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/versioncontrol/GitDiffedCodeExtractionManager.java [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+ * 1) fetch with gitClient necessary files from code repository,
+ * 2) delegate comment extraction to specialized classes and
+ * 3) extract evolution of knowledge elements in files.
+ */
+public class GitDiffedCodeExtractionManager {
+	/*
+	 * [issue]Modified files may add, modify or delete rationale.
+	 * For extraction of rationale on modified files their contents
+	 * from the base and changed versions are required.
+	 * How should both versions of affected files be accessed?{/issue]
```
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/versioncontrol/GitDiffedCodeExtractionManager.java]: Please fix the bracket {/issue].

**Comment by lw2011** [comment]
Hi @kleebaum can I merge this branch with develop? I do not intend to improve it much more in the next days. I need it merged, so I can move forward with CONDEC-503 and visualizations for CONDEC-505.

Cheers
Lukasz

File: src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/CodeCommentWithRange.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package de.uhd.ifi.se.decision.management.jira.extraction;
+
+/** purpose: structure to store comment text
+ * and area it occupies in a source code file
+ */
+public class CodeCommentWithRange {
```
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/CodeCommentWithRange.java]: [alternative] I suggest to simply call this model class "CodeComment"! [/alternative] 
[pro] It is inherent that a CodeComment has a beginning and an end, it does not need to be made explicit in the class' name. [/pro]

File: src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/CodeCommentWithRange.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	public String commentContent = null;
+	// position in source code of the file
+	public int beginColumn = -1;
+	public int beginLine = -1;
```
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/CodeCommentWithRange.java]: What are the column attributes for?

File: src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/RationaleFromCodeCommentExtractor.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * purpose: extract decision knowledge elements from single comment.
+ * Codes rationale source within the source file/comment.
+ */
+public class RationaleFromCodeCommentExtractor {
```
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/RationaleFromCodeCommentExtractor.java]: Great!

File: src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/RationaleFromCodeCommentExtractor.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+				, rationaleType.toUpperCase()
+				, "" // unknown, not needed at the moment
+				, calculateAndCodeRationalePositionInSourceFile(start, rationaleText)
+				, DocumentationLocation.COMMIT);
+	}
+
+	private String calculateAndCodeRationalePositionInSourceFile(int start
+			, String rationaleText) {
+		/**
+		 * @issue: what information the dec. know. key encapsulate
```
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/RationaleFromCodeCommentExtractor.java]: [alternative] I suggest to leave out the colon after the decision knowledge tag! [/alternative] 
[pro] It is not common in JavaDoc to use colons. [/pro]

File: src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/versioncontrol/GitDiffedCodeExtractionManager.java [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+		List<DecisionKnowledgeElement> resultValues = new ArrayList<>();
+
+		if (results.size() > 0) {
+			for (Map.Entry<DiffEntry, CodeExtractionResult> dEntry
+					: results.entrySet()) {
+				String newPath;
+				if (getNew) {
+					newPath = dEntry.getKey().getNewPath();
+				} else {
+					newPath = OLD_FILE_SYMBOL_PREPENDER + dEntry.getKey().getNewPath();
```
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/versioncontrol/GitDiffedCodeExtractionManager.java]: What's the different to getOldPath()?

File: src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/CodeCommentWithRange.java (hunk 3 above)
Comment (lw2011) [reply src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/CodeCommentWithRange.java]: can be removed as it is not used at all

File: src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/versioncontrol/GitDiffedCodeExtractionManager.java (hunk 6 above)
Comment (lw2011) [reply src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/versioncontrol/GitDiffedCodeExtractionManager.java]: oh oh, it should be getOldPath in the else block, thanks!

**Comment by lw2011** [comment]
rebased branch to newest Develop commit.

**Pull request merged**